### ISSUE TITLE
Bot Framework Time Outs and TypeError

### ISSUE THREAD

**Issue by seanbetts** [microsoft/BotFramework-Services#89]
In the last 24 hours I've been having issues with my chatbot that's built on the Bot Framework. I first noticed this when getting Time Out errors on the Facebook channel. Since then I have tested it in Web Chat and locally and I have narrowed the issue down to a Type Error.

<img width="1124" alt="Screenshot 2019-07-13 at 08 09 50" src="https://user-images.githubusercontent.com/19579294/61168401-9ef9ac80-a545-11e9-9fb1-80210284ac53.png">

The chatbot starts up fine and returns a welcome card but beyond that it returns a Type Error for any question you as it. All I'm getting in the console is the following:

[onTurnError]: TypeError: Cannot read property 'map' of undefined

<img width="505" alt="Screenshot 2019-07-13 at 08 07 00" src="https://user-images.githubusercontent.com/19579294/61168371-3d394280-a545-11e9-9077-522f93fbcfe1.png">

I've tested my production code as well as my development code and I'm getting the same error. The production code hasn't been updated for 5 days, so I know it's nothing to do with that.

The only change I've made is to my Azure app service plan where I downgraded it from S1 to F1 as I just came out of my free trial period and I don't need anything more than an F1 environment as I'm still in testing.

I have had a look through my Application Insights and looking at one of the operation_IDs I can see a customEvent and an exception.

In the exception the outerMessage states: POST to MiHNDbot timed out after 15s. Could this be caused by me downgrading my subscription to F1?

**Comment by v-kydela** [comment]
Hello @seanbetts 

It sounds like you're saying the issue occurs even when you test the bot locally, so how could this have to do with downgrading your subscription? Does your bot call any downgraded services that you're hosting in Azure?

If you're testing locally, have you tried debugging the bot in your development environment so you can see the line of code that's throwing the error?

**Comment by seanbetts** [comment]
Hi @v-kydela

Thanks for getting back to me. I think it might be something to do with the QnA Maker service as my bot starts processing a message in the onMessage function but doesn't get beyond calling qnaMaker.

**Comment by v-kydela** [comment]
@seanbetts - So did you try debugging in your development environment?

You can try the QnaMaker API directly to see if you get any errors: https://docs.microsoft.com/en-us/rest/api/cognitiveservices/qnamakerruntime/runtime/generateanswer

**Comment by seanbetts** [comment]
@v-kydela - yes, I've been trying to debug but I'm a bit new to this! It does look like it is QnA Maker as the API isn't working either. I've tried querying it directly and get an error:

"Runtime core is not initialized."

So, it must be that

**Comment by v-kydela** [comment]
@seanbetts 

1. How are you calling the QnAMaker API (cURL etc.)?
2. Can you create a new KB?
3. Can you post screenshots of the problem's you're experiencing?

**Comment by seanbetts** [comment]
@v-kydela - thanks for your help with this.
1. I have tried calling the QnAMakerAPI using cURL
2. I have been able to create a new KB but as soon as I try to save and train the KB I get the same **Bad Argument** error message (see image below)
![4D41CBC8-F617-4942-94D8-FBAF5656C0A9](https://user-images.githubusercontent.com/19579294/61160602-ec4f2d00-a4f7-11e9-8ff9-8a19429e6964.jpeg)
3. I’ll try again in the morning and If I’m still having problems will upload some screenshots

**Comment by seanbetts** [comment]
@v-kydela I'm still having the same problems this morning so have added some screenshots to the posts above. It's obviously a problem with QnA Maker, I just don't know whether it's an issue with my account and KB or whether it's a wider problem, although the status page still says everything is operating normally

**Comment by rokulka** [comment]
@seanbetts This is not a generic QnA Maker service issue. Seems to be with your instance in particular. Could you please create a support ticket and provide your endpoint details so that we could investigate further?

How to create a support ticket -  https://docs.microsoft.com/en-us/azure/azure-supportability/how-to-create-azure-support-request

**Comment by seanbetts** [comment]
Thanks for the reply @rokulka - my endpoint is https://mihndbot-qna.azurewebsites.net/qnamaker

**Comment by seanbetts** [comment]
Thank you everyone for your help - the issue was down to me running out of memory in my app service which I have now resolved and everything is back to working order!

**Comment by diberry** [comment]
@seanbetts - How were you running out of memory? I'm looking for a scenario or something I can add to documentation to help the next person.

**Comment by seanbetts** [comment]
@diberry I had just come out of my free trial period and downgraded my instances as I didn’t need the expensive ones I had set up as default as I am still in the development phase. Without realising I bumped up against the 1Gb limit of the free tier as I had two versions of the chatbot web app set up (prod and dev) at 220Mb each and then my QnA instance at 660Mb each. I would have thought I would have got an alert to tell me I had run out of memory and I wasn’t aware of how much memory I had been using so it’s taken me a few days to figure it out with the help above!

**Comment by diberry** [comment]
@seanbetts The azure web app service for QnA Maker hit a memory consumption boundary without an alert. The result is that QnA Maker fails in several ways (query prediction API, qna portal) Is this accurate? The issue was caused how? automated testing? How did you resolve the issue? Increase pricing tier?

**Comment by seanbetts** [comment]
@diberry Yes, that’s accurate. The issue was caused by me downgrading to the free App Service tier with only 1Gb of memory. It would have been hard to resolve the issue by increasing the pricing tier as both the free and shared development app service tiers only have 1Gb of RAM and so I couldn’t solve the problem unless I upgraded to a full production tier, which I am not ready to do. The way I solved the problem was to create two new app services and move the QnA service and the dev version of the chatbot to these new services, which each have 1Gb of RAM each.
